### PR TITLE
Add librato tag support

### PR DESCRIPTION
**Pull request by twooster**
## What

Add support for Librato-style tags. These appear in the metric name rather than as a separate field, ala:

```
metric.name#tag=val,tag2=val2:0|c
```

## Why

Maybe this is just scratching my itch, but I'm facing the impossibility of attempting to incrementally move from Librato (and librato tags) to an internal Prometheus server. The only way forward for me is to either heavily hack the original Statsd server and the Librato backend plugin to support DogStatsD tags, or to add a bit of support to this exporter, which is what I've opted to do.

## Discussion

This is a fairly straightforward implementation, and is almost entirely backwards compatible, so I didn't bother to invoke the mailing list. The only place this might break is if someone is already using metrics with hash signs in them -- but this should probably not be occurring in most cases. If desired, I can add a configuration flag for this to prevent conflicts.

### CODE REVIEW

**Comment by matthiasr** [comment]
Thank you! This looks reasonable at first glance (TIL that there is another tagging style). I'll go over the code in detail soon.

**Comment by twooster** [comment]
There's unfortunately many different tagging styles. :grinning: There's some discussion here:

https://github.com/statsd/statsd/issues/619

I could relatively simply also support [InfluxDB tagging style](https://www.influxdata.com/blog/getting-started-with-sending-statsd-metrics-to-telegraf-influxdb/#introducing-influx-statsd) (search for either '#' or ',' to begin first tagging), since that's probably also a reasonable and large target to hit? Let me know, and I'll update the PR.

**Comment by matthiasr** [comment]
Fine with me if it's a straightforward change to you!

**Comment by matthiasr** [comment]
OK and the SignalFX way is … [interesting](https://docs.signalfx.com/en/latest/integrations/agent/monitors/collectd-statsd.html#adding-dimensions-to-statsd-metrics) 😂 

I wonder at what point we outgrow if/else and should graduate to some kind of more formalized _parsing_?

**Comment by twooster** [comment]
I've gone ahead and added support for InfluxDB style tags, and updated the documentation and tests to match. :ship:

File: README.md [hunk 1]
```diff
@@ -29,15 +29,52 @@ We recommend this only as an intermediate solution and recommend switching to
...
+See [Tags](https://docs.datadoghq.com/developers/dogstatsd/data_types/#tagging)
+in the DogStatsD documentation for the concept description and
+[Datagram Format](https://docs.datadoghq.com/developers/dogstatsd/datagram_shell/).
+Note that this tagging style is incompatible with the original `statsd`
+implementation, you will be unable to use it as a repeater in front of the
+statsd-to-prometheus exporter.
+
+Although you can use both name-appended tags (Librato or InfluxDB) and
+metric-appended tags simultaneously, this is not recommended. If you do, be
+aware that DogStatsD `name=value` pairs will take priority over Librato tags
```
Comment (matthiasr) [inline README.md]: I suppose they would also take priority over InfluxDB tags?

Comment (twooster) [reply README.md]: Sorry, yup, forgot to change that bit of documentation.

File: bridge_test.go [hunk 2]
```diff
@@ -127,6 +127,46 @@ func TestHandlePacket(t *testing.T) {
...
+			out: Events{
+				&CounterEvent{
+					metricName: "foo",
+					value:      100,
+					labels:     map[string]string{"_09digits": "0", "tag_with_dots": "1"},
+				},
+			},
+		}, {
+			name: "influxdb tag extension",
+			in:   "foo#tag1=bar,tag2=baz:100|c",
```
Comment (matthiasr) [inline bridge_test.go]: this is repeating the librato test, did you mean

```suggestion
			in:   "foo,tag1=bar,tag2=baz:100|c",
```

File: bridge_test.go [hunk 3]
```diff
@@ -127,6 +127,46 @@ func TestHandlePacket(t *testing.T) {
...
+			out: Events{
+				&CounterEvent{
+					metricName: "foo",
+					value:      100,
+					labels:     map[string]string{"tag1": "bar", "tag2": "baz"},
+				},
+			},
+		}, {
+			name: "influxdb tag extension with tag keys unsupported by prometheus",
+			in:   "foo#09digits=0,tag.with.dots=1:100|c",
```
Comment (matthiasr) [inline bridge_test.go]: ditto

```suggestion
			in:   "foo,09digits=0,tag.with.dots=1:100|c",
```

File: exporter.go [hunk 4]
```diff
@@ -343,10 +371,22 @@ func lineToEvents(line string) Events {
...
+	metric := parseNameAndTags(elements[0], labels)
+
 	var samples []string
 	if strings.Contains(elements[1], "|#") {
-		// using datadog extensions, disable multi-metrics
+		// using DogStatsD tags
+
+		// don't allow mixed tagging styles
+		if len(labels) > 0 {
+			sampleErrors.WithLabelValues("malformed_line").Inc()
```
Comment (matthiasr) [inline exporter.go]: Let's count it separate from a generally malformed line:

```suggestion
			sampleErrors.WithLabelValues("mixed_tagging_styles").Inc()
```

File: bridge_test.go (hunk 2 above)
Comment (twooster) [reply bridge_test.go]: Oops. Yes.

File: exporter.go (hunk 4 above)
Comment (twooster) [reply exporter.go]: Works for me.

**Comment by matthiasr** [comment]
Thank you! This is a great addition, and very nicely done.

**Pull request merged**